pkg/downloader/http/downloader.go

```diff
@@ -2,89 +2,126 @@ package http
 
 import (
 	"context"
-	"errors"
+	"fmt"
 	"io"
 	"net/http"
+	"net/url"
 	"os"
-	"time"
+	"path/filepath"
+	"strings"
 
-	"github.com/bacalhau-project/bacalhau/pkg/model"
-	"github.com/bacalhau-project/bacalhau/pkg/system"
+	"github.com/bacalhau-project/bacalhau/pkg/downloader"
+	"github.com/bacalhau-project/bacalhau/pkg/storage/url/urldownload"
 	"github.com/bacalhau-project/bacalhau/pkg/util/closer"
 	"github.com/rs/zerolog/log"
 )
 
+// Replace slashes with some other character that is valid for filenames in most operating systems
+var urlSanitizer = strings.NewReplacer("/", "_", "\\", "_", ":", "_", "*", "_", "?", "_", "\"", "_", "<", "_", ">", "_", "|", "_")
+
 type Downloader struct {
-	Settings *model.DownloaderSettings
+	httpClient *http.Client
 }
 
-func NewHTTPDownloader(settings *model.DownloaderSettings) *Downloader {
-	return &Downloader{
-		Settings: settings,
-	}
+// NewHTTPDownloader creates a new HTTPDownloader with the given settings.
+func NewHTTPDownloader() *Downloader {
+	return &Downloader{httpClient: http.DefaultClient}
 }
 
-func (httpDownloader *Downloader) IsInstalled(context.Context) (bool, error) {
+// IsInstalled checks if the downloader is ready to be used.
+func (httpDownloader *Downloader) IsInstalled(ctx context.Context) (bool, error) {
+	// For HTTPDownloader, we can always return true as there's no installation needed.
 	return true, nil
 }
 
-func (httpDownloader *Downloader) DescribeResult(ctx context.Context, result model.PublishedResult) (map[string]string, error) {
-	return nil, errors.New("not implemented for httpdownloader")
-}
-
-func (httpDownloader *Downloader) FetchResult(ctx context.Context, item model.DownloadItem) error {
-	ctx, span := system.NewSpan(ctx, system.GetTracer(), "pkg/downloader/http.Downloader.FetchResults")
-	defer span.End()
-
-	err := func() error {
-		log.Ctx(ctx).Debug().Msgf(
-			"Downloading result URL %s '%s' to '%s'...",
-			item.Name,
-			item.CID, item.Target,
-		)
-
-		innerCtx, cancel := context.WithDeadline(ctx, time.Now().Add(httpDownloader.Settings.Timeout))
-		defer cancel()
-
-		return fetch(innerCtx, item.CID, item.Target)
-	}()
+// FetchResult downloads the result of a computation and saves it to a local file.
+func (httpDownloader *Downloader) FetchResult(ctx context.Context, item downloader.DownloadItem) (string, error) {
```

Comment:
I know it's not the time to do it, but this whole downloader package feels like it should be integrated with the publishing for the storage providers.